### DISCUSSION TITLE
Support for ExceptionGroup

### DISCUSSION THREAD

**Discussion by nameer** [fastapi/fastapi#9826] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel, validator

from string import punctuation


app = FastAPI()


class Data(BaseModel):
    password: str

    @validator("password")
    def v_pwd(cls, a):
        e = []
        if not any(char.islower() for char in a):
            e.append(ValueError("Atleast one lowercase char is required"))
        if not any(char.isupper() for char in a):
            e.append(ValueError("Atleast one uppercase char is required"))
        if not any(char.isdigit() for char in a):
            e.append(ValueError("Atleast one digit is required"))
        if not any(char in punctuation for char in a):
            e.append(ValueError("Atleast one special character is required"))
        if e:
            raise ExceptionGroup("Password strength", e)
        return a


@app.post("/eg")
async def exception_group(data: Data) -> str:
    return "Success"
```

### Description

ExceptionGroup is supported starting from Python 3.11, but FastAPI does not currently provide built-in handling for ExceptionGroup.

As shown in the code, it would be advantageous if FastAPI incorporated a mechanism to handle ExceptionGroup, particularly for exceptions that can be managed within the FastAPI framework.

### Operating System

Linux

### Operating System Details

Ubuntu 22.04.2

### FastAPI Version

0.99.1

### Python Version

3.11.4

**Comment by YuriiMotov** [comment, marked as answer]
Could you please clarify your expectations?
It was probably fixed since that time. The following code works with FastAPI 0.115.12, Pydantic 2.11.4:

```python
from fastapi import FastAPI
from fastapi.responses import JSONResponse
from pydantic import BaseModel, field_validator

from string import punctuation


app = FastAPI()


class Data(BaseModel):
    password: str

    @field_validator("password")
    @classmethod
    def v_pwd(cls, a):
        e = []
        if not any(char.islower() for char in a):
            e.append(ValueError("Atleast one lowercase char is required"))
        if not any(char.isupper() for char in a):
            e.append(ValueError("Atleast one uppercase char is required"))
        if not any(char.isdigit() for char in a):
            e.append(ValueError("Atleast one digit is required"))
        if not any(char in punctuation for char in a):
            e.append(ValueError("Atleast one special character is required"))
        if e:
            raise ExceptionGroup("Password strength", e)
        return a


@app.post("/eg")
async def exception_group(data: Data) -> str:
    return "Success"


@app.exception_handler(ExceptionGroup)
async def exception_group_handler(request, exc: ExceptionGroup):
    return JSONResponse(
        status_code=422,
        content={
            "message": "Password is not strong enough",
            "errors": [{"error": str(e)} for e in exc.exceptions],
        },
    )

```

Results:

```
curl -X 'POST' \
  'http://localhost:8000/eg' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"password": "string"}'

{"message":"Password is not strong enough","errors":[{"error":"Atleast one uppercase char is required"},{"error":"Atleast one digit is required"},{"error":"Atleast one special character is required"}]}

```

BTW, thanks for pointing to this feature. Very interesting!

Can we now close this as resolved?

**Reply by mikenerone** [reply]
Yes, of course `ExceptionGroup` is itself an exception - in that sense it was supported: you could write a handler for it. But this is not a particularly useful pattern for an `ExceptionGroup`. In your example, it is assumed that the `ExceptionGroup` corresponds to a known password strength problem, but that's rarely the case. Proper support would involve defining some semantics for selecting a handler based on the exceptions actually contained within the exception group. I think something like this is probably what the OP originally had in mind.

What that would look like isn't a simple topic, and I don't personally have a design in mind to suggest. Some high-level options:
1. Just find the first contained exception and proceed as if _that_ was the raised exception. This loses the context of all of the other exceptions.
2. Find the first exception that _does_ have a handler, and delegate to that. If none, then proceed as if no handler. This still loses the other exceptions, but at least gets more opportunities to find a meaningful handler to construct a response.
3. More correct, but unless someone has a very clever idea, quite a bit more complicated: design a completely new way of handling exception groups that preserves all of the contained exceptions. All corresponding handlers are called, and each, _somehow_, contributes necessary details to the construction of a single response that in some way reflects all of the errors. No simple way to achieve this comes immediately to mind.

**Reply by YuriiMotov** [reply]
Haven't tried it, but what if we create custom exception group and raise it, then handle it using exception handler?

```py

from fastapi import FastAPI
from fastapi.responses import JSONResponse
from pydantic import BaseModel, field_validator

from string import punctuation

class PasswordIssue(ExceptionGroup):
    pass


app = FastAPI()


class Data(BaseModel):
    password: str

    @field_validator("password")
    @classmethod
    def v_pwd(cls, a):
        e = []
        if not any(char.islower() for char in a):
            e.append(ValueError("Atleast one lowercase char is required"))
        if not any(char.isupper() for char in a):
            e.append(ValueError("Atleast one uppercase char is required"))
        if not any(char.isdigit() for char in a):
            e.append(ValueError("Atleast one digit is required"))
        if not any(char in punctuation for char in a):
            e.append(ValueError("Atleast one special character is required"))
        if e:
            raise PasswordIssue("Password strength", e)
        return a


@app.post("/eg")
async def exception_group(data: Data) -> str:
    return "Success"


@app.exception_handler(PasswordIssue)
async def exception_group_handler(request, exc: PasswordIssue):
    return JSONResponse(
        status_code=422,
        content={
            "message": "Password is not strong enough",
            "errors": [{"error": str(e)} for e in exc.exceptions],
        },
    )

```
